### DISCUSSION TITLE
How to execute Boolean Operation on two 3D Mesh ?

### DISCUSSION THREAD

**Discussion by ankofl** [CGAL/cgal#8306] [Q&A] [answered]
I have four arrays: 

Point3[] pointsOne, 
Point3[] pointsTwo, 
int[] IndexesOne,
int[] IndexesTwo

How can I perform the Boolean operation of subtracting one Mesh from another?

**Comment by lrineau** [comment]
Your question is a bit vague. That depends on the way you mesh is structured. Is it a mesh of triangles (all faces have exactly three vertices), or more general?

In CGAL, that sort of data is a called a polygon soup. We have functions to process that, in the chapter [Polygon Mesh Processing] of the documentation.

[Polygon Mesh Processing]: https://doc.cgal.org/latest/Polygon_mesh_processing/index.html

**Reply by ankofl** [reply]
Thank you for the quick response! 

Each polygon has exactly three vertices, the normal of almost every polygon is correctly oriented inside or outside the body volume (although if there is a function that allows validating and correcting input data, this will be a huge plus!)

I have already seen this example in the documentation:
```c
PMP::corefine_and_compute_union(mesh1,mesh2, out)
````

but it contains the creation of a Mesh from an off file, but is there a similar method that creates a Mesh from an array of vectors?
```c
PMP::IO::read_polygon_mesh("blobby.off", mesh1)
PMP::IO::read_polygon_mesh("eight.off", mesh2)
```

It would also be great if there is an analog method 
```c
CGAL::IO::write_polygon_mesh("union.off", out, CGAL::parameters::stream_precision(17))
```
for converting Mesh back to an array of triangles or two arrays (Point3[] points and int[] indices)

P.S. Reading and writing files creates a lot of overhead, and considering that I already have all the data uploaded, I would like to get rid of the overhead

Thanks!

**Reply by MaelRL** [reply, marked as answer]
The following functions can be used to go from a soup (ranges of points and indices) to a mesh, and back:
- [PMP::polygon_soup_to_polygon_mesh()](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__combinatorial__repair__grp.html#ga2ba9722ec8472a1455107ffce7145e46)
- [PMP::polygon_mesh_to_polygon_soup()](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__combinatorial__repair__grp.html#ga3be94f1f7968022a294dd533400ad554)

The function [PMP::orient_polygon_soup()](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__orientation__grp.html#gad380465ee62d858d27fab4cfda6c1764) could help with your badly oriented facets.

See the example: https://doc.cgal.org/latest/Polygon_mesh_processing/Polygon_mesh_processing_2repair_polygon_soup_example_8cpp-example.html for a combined use.

**Reply by ankofl** [reply]
Thank you very much!
